File: 200-number-of-islands/number-of-islands.java
class Solution {
    public int numIslands(char[][] grid) {
        int m = grid.length;
        int n = grid[0].length;
        int[][] visited = new int[m][n];
        int num_islands = 0;

        for (int i = 0; i < m; i++) {
            for (int j = 0; j < n; j++) {
                if (grid[i][j] == '1' && visited[i][j] == 0) {
                    num_islands++;
                    dfs(grid, i, j, visited);
                }
            }
        }
        return num_islands;
    }

    private void dfs(char[][] grid, int i, int j, int[][] visited) {
        // index exceeds bounds or already visited 
        if (i >= grid.length || j >= grid[0].length || i < 0 || j < 0
        || visited[i][j] == 1 || grid[i][j] == '0') {
            return;
        }
        visited[i][j] = 1;

        dfs(grid, i + 1, j, visited); 
        dfs(grid, i - 1, j, visited);
        dfs(grid, i, j + 1, visited); 
        dfs(grid, i, j - 1, visited);
    }
}

File: 200-number-of-islands/number-of-islands.py
class Solution:
    def numIslands(self, grid: List[List[str]]) -> int:
        # island -> 1's surrounded by 0's
        res = 0
        m = len(grid)
        n = len(grid[0])

        def helper(r, c):
            if not (0 <= r < m) or not (0 <= c < n) or grid[r][c] == "0":
                return 

            grid[r][c] = "0"

            helper(r - 1, c)
            helper(r + 1, c)
            helper(r, c - 1)
            helper(r, c + 1)

        for r in range(m):
            for c in range(n):
                if grid[r][c] == "1":
                    res += 1
                    # mark island group (all reachable 1's) as visited
                    helper(r, c)

        return res